hpx/runtime/components/default_distribution_policy.hpp

```diff
@@ -271,63 +274,85 @@ namespace hpx { namespace components
         ///
         std::size_t get_num_localities() const
         {
-            return (std::max)(std::size_t(1), localities_.size());
+            return (std::max)(std::size_t(1), localities_->size());
         }
 
         /// Returns the locality which is anticipated to be used for the next
         /// async operation
         hpx::id_type get_next_target() const
         {
-            return localities_.empty() ? hpx::find_here() : localities_.front();
+            return (!localities_ || localities_->empty()) ?
+                hpx::find_here() : localities_->front();
         }
 
     protected:
         /// \cond NOINTERNAL
-        std::size_t
-        get_num_items(std::size_t items, hpx::id_type const& loc) const
+        std::size_t get_num_items(
+            std::size_t items, hpx::id_type const& loc) const
         {
             // make sure the given id is known to this distribution policy
             HPX_ASSERT(
-                std::find(localities_.begin(), localities_.end(), loc) !=
-                    localities_.end() ||
-                (localities_.empty() && loc == hpx::find_here())
+               !localities_ ||
+                (localities_->empty() && loc == hpx::find_here()) ||
+                std::find(localities_->begin(), localities_->end(), loc) !=
+                    localities_->end()
             );
 
+            if (!localities_ || localities_->empty())
+            {
+                return std::size_t(1);
+            }
+
             // this distribution policy places an equal number of items onto
             // each locality
-            std::size_t locs = (std::max)(std::size_t(1), localities_.size());
+            std::size_t locs = (std::max)(std::size_t(1), localities_->size());
 
             // the overall number of items to create is smaller than the number
             // of localities
             if (items < locs)
             {
-                auto it = std::find(localities_.begin(), localities_.end(), loc);
-                std::size_t num_loc = std::distance(localities_.begin(), it);
+                auto it = std::find(localities_->begin(), localities_->end(), loc);
+                std::size_t num_loc = std::distance(localities_->begin(), it);
                 return (items < num_loc) ? 1 : 0;
             }
 
             // the last locality might get less items
-            if (localities_.size() > 1 && loc == localities_.back())
+            if (localities_->size() > 1 && loc == localities_->back())
+            {
                 return items - detail::round_to_multiple(items, locs, locs-1);
+            }
 
             // otherwise just distribute evenly
             return (items + locs - 1) / locs;
         }
+
+        std::shared_ptr<std::vector<id_type>> ensure_localities() const
```

Comment:
This ensure localities is used inconsistently throughout the patch. Why not just create a shared pointer with an empyt vector in the default ctor?

Reply:
What would that change? We would have to delay-fill it anyways

Reply:
What changes is that we get rid of the checks if `localities_` is not pointing to a `nullptr`. Since we require at least a single element vector, this seems like the perfect fit for having in a default ctor, no?

Reply:
We can't fill the vector with `hpx::find_here()` in the default constructor as this may cause HPX functionality to be invoked during startup. That means we would have to check whether the vector is empty anyways. I don't really see a benefit of what you suggest.

Reply:
Ok. I was confused since we didn't need this before. There are various uses of `localities_` directly. There doesn't seem to be a consistent way of when to ensure that `localities_` exists or not. I think to resemble the old behavior, it would make sense to at least have a `std::make_shared<std::vector<hpx::id_type>()` in the default ctor. Bringing back the old functionality without introducing further complexity.
Or change the implementation such that an empty `shared_ptr` is equivalent to the previous behavior of an empty vector. Right now, the two ways of expressing the same thing lead to confusion.

Reply:
Could you elaborate what you mean by 'two ways of expressing the same thing', please? The use of the `localities_` vector is well defined: it's either pre-defined during construction or it's populated on demand whenever accessed. 

Yes, this adds more functionality than we've initially had: it now allows to adjust the policy to a changing set of available localities. Is that a bad thing?

Reply:
`localities_` essentially has three states:

1. `localities_ == nullptr`
2. `localities_->size() == 0`
3. `localities_->size() > 0`

1. and 2. are currently more or less equivalent. However, some places in the code use `ensure_localities()`, and others don't but check for `localities_ != nullptr`explicitly (for example here: https://github.com/STEllAR-GROUP/hpx/pull/3087/files#diff-61639ba7144f3b99ed04835f11dd7924R301). And other places use `localities_` without a `nullptr` check (for example here: https://github.com/STEllAR-GROUP/hpx/pull/3087/files#diff-61639ba7144f3b99ed04835f11dd7924R277).

I don't see how we are now able to have a changing set of localities. As before, we can create policies on demand with changing sets of localities.

Reply:
Well, `localities_->size() == 0` is essentially an error condition. I'd be willing to rule that case out.

Reply:
> and others don't but check for localities_ != nullptrexplicitly (for example here: https://github.com/STEllAR-GROUP/hpx/pull/3087/files#diff-61639ba7144f3b99ed04835f11dd7924R301)

Well, we could use `ensure_localities()` in this case as well. I can change that.

> And other places use localities_ without a nullptr check (for example here: https://github.com/STEllAR-GROUP/hpx/pull/3087/files#diff-61639ba7144f3b99ed04835f11dd7924R277).

Right, this is a bug, will fix.

Reply:
What I am suggesting is to keep the "empty vector" case and prefer it over the empty shared pointer. The shared pointer is needed for the asynchronous dataflow operation. The empty vector means "here". In all cases, a `shared_ptr<vector<id_type>>` is being constructed. If we are in the "default" constructed case, we'll create a new shared pointer and a new one element vector on each access. I don't think that's what we want.

Reply:
I don't think this is a good idea. It is completely useless to pre-allocate the shared_ptr with an empty vector in the default-constructed case. The (empty) vector can't be used for anything. We would have to delay-fill that vector later on in order to make it useful. This however may turn out to be a thread-safety problem (would require a lock).  

OTOH, if we left the shared_ptr empty in the default-constructed case we would create a new (independent) shared_ptr on demand and fill it as needed (as suggested by this PR). No lock is needed.